// next-sanity/dist/SanityLive.js
import { d as setEnvironment, f as setPerspective } from "./context.js";
import { t as isCorsOriginError } from "./isCorsOriginError.js";
import { useRouter } from "next/navigation";
import { startTransition, useEffect, useEffectEvent, useMemo, useRef, useState } from "react";
import { isMaybePresentation, isMaybePreviewWindow } from "@sanity/presentation-comlink";
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import { createClient } from "@sanity/client";
import dynamic from "next/dynamic";
import { revalidateSyncTags } from "next-sanity/live/server-actions";
const PresentationComlink = dynamic(() => import("./PresentationComlink.js"));
const RefreshOnFocus = dynamic(() => import("./RefreshOnFocus.js"));
const RefreshOnMount = dynamic(() => import("./RefreshOnMount.js"));
const RefreshOnInterval = dynamic(() => import("./RefreshOnInterval.js"));
const RefreshOnReconnect = dynamic(() => import("./RefreshOnReconnect.js"));
function SanityLive(props) {
	const { projectId, dataset, apiHost, apiVersion, useProjectHostname, token, requestTagPrefix, draftModeEnabled, draftModePerspective, requestTag = "next-loader.live", waitFor, revalidateSyncTags: revalidateSyncTags$1 = revalidateSyncTags, onError = handleError, intervalOnGoAway = 3e4, onGoAway = handleOnGoAway, refreshOnMount = false, refreshOnFocus = draftModeEnabled ? false : typeof window === "undefined" ? true : window.self === window.top, refreshOnReconnect = true } = props;
	const client = useMemo(() => createClient({
		projectId,
		dataset,
		apiHost,
		apiVersion,
		useProjectHostname,
		ignoreBrowserTokenWarning: true,
		token,
		useCdn: false,
		requestTagPrefix
	}), [
		apiHost,
		apiVersion,
		dataset,
		projectId,
		requestTagPrefix,
		token,
		useProjectHostname
	]);
	const [refreshOnInterval, setRefreshOnInterval] = useState(false);
	/**
	* 1. Handle Live Events and call revalidateTag or router.refresh when needed
	*/
	const router = useRouter();
	const handleLiveEvent = useEffectEvent((event) => {
		if (process.env.NODE_ENV !== "production" && event.type === "welcome") {
			console.info("Sanity is live with", token ? "automatic revalidation for draft content changes as well as published content" : draftModeEnabled ? "automatic revalidation for only published content. Provide a `browserToken` to `defineLive` to support draft content outside of Presentation Tool." : "automatic revalidation of published content");
			startTransition(() => setRefreshOnInterval(false));
		} else if (event.type === "message") if (waitFor === "function") startTransition(() => router.refresh());
		else revalidateSyncTags$1(event.tags).then((result) => {
			if (result === "refresh") startTransition(() => router.refresh());
		});
		else if (event.type === "restart" || event.type === "reconnect") {
			startTransition(() => setRefreshOnInterval(false));
			startTransition(() => router.refresh());
		} else if (event.type === "goaway") {
			onGoAway(event, intervalOnGoAway);
			startTransition(() => setRefreshOnInterval(intervalOnGoAway));
		}
	});
	useEffect(() => {
		const subscription = client.live.events({
			includeDrafts: !!token,
			tag: requestTag,
			waitFor
		}).subscribe({
			next: handleLiveEvent,
			error: (err) => {
				onError(err);
			}
		});
		return () => subscription.unsubscribe();
	}, [
		client.live,
		onError,
		requestTag,
		token,
		waitFor
	]);
	/**
	* 2. Notify what perspective we're in, when in Draft Mode
	*/
	useEffect(() => {
		if (draftModeEnabled && draftModePerspective) setPerspective(draftModePerspective);
		else setPerspective("unknown");
	}, [draftModeEnabled, draftModePerspective]);
	const [loadComlink, setLoadComlink] = useState(false);
	/**
	* 3. Notify what environment we're in, when in Draft Mode
	*/
	useEffect(() => {
		if (isMaybePresentation()) return;
		if (draftModeEnabled && token) {
			setEnvironment("live");
			return;
		}
		if (draftModeEnabled) {
			setEnvironment("static");
			return;
		}
		setEnvironment("unknown");
	}, [draftModeEnabled, token]);
	/**
	* 4. If Presentation Tool is detected, load up the comlink and integrate with it
	*/
	useEffect(() => {
		if (!isMaybePresentation()) return;
		const controller = new AbortController();
		const timeout = setTimeout(() => setEnvironment("live"), 3e3);
		window.addEventListener("message", ({ data }) => {
			if (data && typeof data === "object" && "domain" in data && data.domain === "sanity/channels" && "from" in data && data.from === "presentation") {
				clearTimeout(timeout);
				setEnvironment(isMaybePreviewWindow() ? "presentation-window" : "presentation-iframe");
				setLoadComlink(true);
				controller.abort();
			}
		}, { signal: controller.signal });
		return () => {
			clearTimeout(timeout);
			controller.abort();
		};
	}, []);
	/**
	* 5. Warn if draft mode is being disabled
	* @TODO move logic into PresentationComlink, or maybe VisualEditing?
	*/
	const draftModeEnabledWarnRef = useRef(void 0);
	useEffect(() => {
		if (!draftModeEnabled) return;
		clearTimeout(draftModeEnabledWarnRef.current);
		return () => {
			draftModeEnabledWarnRef.current = setTimeout(() => {
				console.warn("Sanity Live: Draft mode was enabled, but is now being disabled");
			});
		};
	}, [draftModeEnabled]);
	return /* @__PURE__ */ jsxs(Fragment, { children: [
		draftModeEnabled && loadComlink && /* @__PURE__ */ jsx(PresentationComlink, {
			projectId,
			dataset,
			draftModeEnabled,
			draftModePerspective
		}),
		!draftModeEnabled && refreshOnMount && /* @__PURE__ */ jsx(RefreshOnMount, {}),
		refreshOnInterval && Number.isFinite(refreshOnInterval) && refreshOnInterval > 0 && /* @__PURE__ */ jsx(RefreshOnInterval, { interval: refreshOnInterval }),
		!draftModeEnabled && refreshOnFocus && /* @__PURE__ */ jsx(RefreshOnFocus, {}),
		!draftModeEnabled && refreshOnReconnect && /* @__PURE__ */ jsx(RefreshOnReconnect, {})
	] });
}
SanityLive.displayName = "SanityLiveClientComponent";
function handleError(error) {
	if (isCorsOriginError(error)) console.warn(`Sanity Live is unable to connect to the Sanity API as the current origin - ${window.origin} - is not in the list of allowed CORS origins for this Sanity Project.`, error.addOriginUrl && `Add it here:`, error.addOriginUrl?.toString());
	else console.error(error);
}
function handleOnGoAway(event, intervalOnGoAway) {
	if (intervalOnGoAway) console.warn("Sanity Live connection closed, switching to long polling set to a interval of", intervalOnGoAway / 1e3, "seconds and the server gave this reason:", event.reason);
	else console.error("Sanity Live connection closed, automatic revalidation is disabled, the server gave this reason:", event.reason);
}
export { SanityLive as default };
