// next-sanity/dist/hooks/index.js
"use client";
import { a as environment, c as perspectiveListeners, i as comlinkProjectId, n as comlinkDataset, o as environmentListeners, r as comlinkListeners, s as perspective, t as comlink } from "../context.js";
import { useCallback, useEffect, useEffectEvent, useMemo, useReducer, useSyncExternalStore } from "react";
import { stegaEncodeSourceMap } from "@sanity/client/stega";
import { dequal } from "dequal/lite";
import { useIsPresentationTool, useOptimistic, useVisualEditingEnvironment } from "@sanity/visual-editing/react";
/**
* Reports the current draft mode environment.
* Use it to determine how to adapt the UI based on wether:
* - Your app is previewed in a iframe, inside Presentation Tool in a Sanity Studio.
* - Your app is previewed in a new window, spawned from Presentation Tool in a Sanity Studio.
* - Your app is live previewing drafts in a standalone context.
* - Your app is previewing drafts, but not live.
* - Your app is not previewing anything (that could be detected).
* @public
* @deprecated this hook will be removed in the next major version
*/
function useDraftModeEnvironment() {
	return useSyncExternalStore(useCallback((listener) => {
		environmentListeners.add(listener);
		return () => environmentListeners.delete(listener);
	}, []), () => environment, () => "checking");
}
/**
* Reports the Sanity Client perspective used to fetch data in `sanityFetch` used on the page.
* If the hook is used outside Draft Mode it will resolve to `'unknown'`.
* If the hook is used but the `<SanityLive />` component is not present then it'll stay in `'checking'` and console warn after a timeout that it seems like you're missing the component.
* @public
* @deprecated this hook will be removed in the next major version
*/
function useDraftModePerspective() {
	return useSyncExternalStore(useCallback((listener) => {
		perspectiveListeners.add(listener);
		return () => perspectiveListeners.delete(listener);
	}, []), () => perspective, () => "checking");
}
/**
* Detects if the application is considered to be in a "Live Preview" mode.
* Live Preview means that the application is either:
* - being previewed inside Sanity Presentation Tool
* - being previewed in Draft Mode, with a `browserToken` given to `defineLive`, also known as "Standalone Live Preview'"
* When in Live Preview mode, you typically want UI to update as new content comes in, without any manual intervention.
* This is very different from Live Production mode, where you usually want to delay updates that might cause layout shifts,
* to avoid interrupting the user that is consuming your content.
* This hook lets you adapt to this difference, making sure production doesn't cause layout shifts that worsen the UX,
* while in Live Preview mode layout shift is less of an issue and it's better for the editorial experience to auto refresh in real time.
*
* The hook returns `null` initially, to signal it doesn't yet know if it's live previewing or not.
* Then `true` if it is, and `false` otherwise.
* @public
* @deprecated this hook will be removed in the next major version
*/
function useIsLivePreview() {
	const environment = useDraftModeEnvironment();
	return environment === "checking" ? null : environment === "presentation-iframe" || environment === "presentation-window" || environment === "live";
}
function reducer(state, { type, payload }) {
	switch (type) {
		case "query-change": return dequal(state, payload) ? state : {
			...state,
			data: dequal(state.data, payload.data) ? state.data : payload.data,
			sourceMap: dequal(state.sourceMap, payload.sourceMap) ? state.sourceMap : payload.sourceMap,
			perspective: dequal(state.perspective, payload.perspective) ? state.perspective : payload.perspective
		};
		default: return state;
	}
}
const initialState = {
	data: null,
	sourceMap: null,
	perspective: null
};
function subscribe(listener) {
	comlinkListeners.add(listener);
	return () => comlinkListeners.delete(listener);
}
const EMPTY_QUERY_PARAMS = {};
const LISTEN_HEARTBEAT_INTERVAL = 1e4;
/**
* Experimental hook that can run queries in Presentation Tool.
* Query results are sent back over postMessage whenever the query results change.
* It also works with optimistic updates in the studio itself, offering low latency updates.
* It's not as low latency as the `useOptimistic` hook, but it's a good compromise for some use cases.
* Especially until `useOptimistic` propagates edits in the Studio parent window back into the iframe.
* @alpha
*/
function usePresentationQuery(props) {
	const [state, dispatch] = useReducer(reducer, initialState);
	const { query, params = EMPTY_QUERY_PARAMS, stega = true } = props;
	/**
	* Comlink forwards queries we want to run to the parent window where Presentation Tool handles it for us
	*/
	const comlink$1 = useSyncExternalStore(subscribe, () => comlink, () => null);
	/**
	* The comlink events requires projectId and dataset, Presentation Tool uses it to protect against project and dataset mismatch errors.
	* We don't want to force the consumers of the `usePresentationQuery` hook to provide these,
	* so we set them in the component that establishes the comlink connection and propagates it to all the subscribes.
	*/
	const projectId = useSyncExternalStore(subscribe, () => comlinkProjectId, () => null);
	const dataset = useSyncExternalStore(subscribe, () => comlinkDataset, () => null);
	/**
	* The perspective is kept in sync with Presentation Tool's perspective, and even knows what perspective the page loaded with initially and can forward it to the Sanity Studio.
	*/
	const perspective = useDraftModePerspective();
	const handleQueryHeartbeat = useEffectEvent((comlink) => {
		if (!projectId || !dataset || !perspective) {
			console.warn("usePresentationQuery: projectId, dataset and perspective must be set", {
				projectId,
				dataset,
				perspective
			});
			return;
		}
		if (perspective === "checking" || perspective === "unknown") return;
		comlink.post("loader/query-listen", {
			projectId,
			dataset,
			perspective,
			query,
			params,
			heartbeat: LISTEN_HEARTBEAT_INTERVAL
		});
	});
	const handleQueryChange = useEffectEvent((event) => {
		if (dequal({
			projectId,
			dataset,
			query,
			params
		}, {
			projectId: event.projectId,
			dataset: event.dataset,
			query: event.query,
			params: event.params
		})) dispatch({
			type: "query-change",
			payload: {
				data: event.result,
				sourceMap: event.resultSourceMap || null,
				perspective: event.perspective
			}
		});
	});
	useEffect(() => {
		if (!comlink$1) return;
		const unsubscribe = comlink$1.on("loader/query-change", handleQueryChange);
		const interval = setInterval(() => handleQueryHeartbeat(comlink$1), LISTEN_HEARTBEAT_INTERVAL);
		return () => {
			clearInterval(interval);
			unsubscribe();
		};
	}, [comlink$1]);
	return useMemo(() => {
		if (stega && state.sourceMap) return {
			...state,
			data: stegaEncodeSourceMap(state.data, state.sourceMap, {
				enabled: true,
				studioUrl: "/"
			})
		};
		return state;
	}, [state, stega]);
}
export { useDraftModeEnvironment, useDraftModePerspective, useIsLivePreview, useIsPresentationTool, useOptimistic, usePresentationQuery, useVisualEditingEnvironment };
