// next-sanity/dist/live/server-actions/index.js
"use server";
import { t as sanitizePerspective } from "../../sanitizePerspective.js";
import { perspectiveCookieName } from "@sanity/preview-url-secret/constants";
import { cookies, draftMode } from "next/headers";
import { revalidateTag, updateTag } from "next/cache";
/**
* @internal CAUTION: this is an internal action and does not follow semver. Using it directly is at your own risk.
*/
async function revalidateSyncTags(tags) {
	const { isEnabled: isDraftMode } = await draftMode();
	if (!isDraftMode) revalidateTag("sanity:fetch-sync-tags", "max");
	const logTags = [];
	for (const _tag of tags) {
		const tag = `sanity:${_tag}`;
		if (isDraftMode) revalidateTag(tag, "max");
		else updateTag(tag);
		logTags.push(tag);
	}
	console.log(`<SanityLive /> ${isDraftMode ? `revalidated tags: ${logTags.join(", ")} with cache profile "max" ` : `updated tags: ${logTags.join(", ")} and revalidated tag: "sanity:fetch-sync-tags" with cache profile "max"`}`);
	if (isDraftMode) return "refresh";
}
/**
* @internal CAUTION: this is an internal action and does not follow semver. Using it directly is at your own risk.
*/
async function setPerspectiveCookie(perspective) {
	if (!(await draftMode()).isEnabled) return;
	const sanitizedPerspective = sanitizePerspective(perspective, "drafts");
	if (perspective !== sanitizedPerspective) throw new Error(`Invalid perspective`, { cause: perspective });
	(await cookies()).set(perspectiveCookieName, Array.isArray(sanitizedPerspective) ? sanitizedPerspective.join(",") : sanitizedPerspective, {
		httpOnly: true,
		path: "/",
		secure: true,
		sameSite: "none"
	});
}
export { revalidateSyncTags, setPerspectiveCookie };
